src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,162 @@
+    }
+
     @Override
     public void add(T value) {
+        addLast(value);
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexToAdd(index);
+        if (index == size) {
+            addLast(value);
+        } else {
+            addBefore(value, getNodeByIndex(index));
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        if (list != null) {
+            for (T value : list) {
+                add(value);
+            }
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return getNodeByIndex(index).element;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNodeByIndex(index);
+        T oldValue = current.element;
+        current.element = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = getNodeByIndex(index);
+        unlink(current);
+        size--;
+        return current.element;
     }
 
     @Override
     public boolean remove(T object) {
-        return false;
+        int index = getNodeIndexByValue(object);
+        if (index == -1) {
+            return false;
+        }
+        remove(index);
+        return true;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void addLast(T value) {
+        Node<T> newNode = new Node<>(null, value, null);
+        if (size == 0) {
+            first = last = newNode;
+        } else {
+            last.next = newNode;
+            newNode.prev = last;
+            last = newNode;
+        }
+    }
+
+    private void addBefore(T value, Node<T> current) {
+        Node<T> previous = current.prev;
+        Node<T> newNode = new Node<>(previous, value, current);
+        current.prev = newNode;
+        if (previous == null) {
+            first = newNode;
+        } else {
+            previous.next = newNode;
+        }
+    }
+
+    private void checkIndexToAdd(int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index " + index
+                    + " is out of bounds for size " + size);
+        }
+    }
+
+    private void checkIndex(int index) {
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index " + index
+                    + " is out of bounds for size " + size);
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        Node<T> current = first;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current;
+    }
+
+    private int getNodeIndexByValue(T value) {
```

Comment:
```suggestion
    private Node<T> getNodeByValue(T value) {
```
this type of list better return node, becaose letter you egain do search node line 78